Add the ability to disable/enable the context menu

loopmode opened the pull request:
This PR adds a dispose function and fixes https://github.com/sindresorhus/electron-context-menu/issues/82 and fixes https://github.com/sindresorhus/electron-context-menu/issues/84.

The implementation is such that the `contextMenu()` return value is a function that will remove the registered event listeners.

loopmode commented:
@sindresorhus I also have updated the fixture.js and html files, but I'm not sure whether I did it "correctly", so it's parked on a separate branch. See https://github.com/loopmode/electron-context-menu/commit/26f36c4e6d9a2d709b57dfcbdee038d3fe190675

It adds a checkbox to enable or disable the context menu, but for that to work while keeping it simple, I moved the entire logic our of the js and into the html file.

I can provide another PR for that.

sindresorhus commented on index.js:
```diff
@@ -244,29 +244,51 @@ const create = (win, options) => {
+	};
+
+	webContents(win).on('context-menu', onContextMenu);
+	return () => webContents(win).removeListener('context-menu', onContextMenu);
```
I prefer not using an inline arrow function when there's no return value. It just makes the code harder to read. Applies in other places too.

loopmode replied:
Returning a named `function dispose() { ...` is clearly the better option, also when debugging etc.

sindresorhus commented on index.js:
```diff
@@ -244,29 +244,51 @@ const create = (win, options) => {
+
+	const dispose = () => {
+		disposables.forEach(dispose => dispose());
+		isDisposed = true;
```
Instead of a boolean, just empty the array and check the length. You should empty the array regardless as you don't want to hold onto those forever.

loopmode replied:
Hmm. I believe it would cause a chicken/egg problem and we'd never create the menu in the first place. Looking into it.

loopmode replied:
Indeed, it's not that easy. 
Avoiding the flag requires either a completely different approach, or several more changes that seem unnecessarily complex.
It's really about bailing out in certain deferred edge cases, and the significant piece of information of the flag is: "No need to create the menu, the callsite already disposed it".

I already tried a couple variants, and I still believe the flag is the best solution. 

Should let go of the references despite of that.

loopmode replied:
array length reset

sindresorhus commented:
This looks great :)

You also need to add it to https://github.com/sindresorhus/electron-context-menu/blob/master/index.d.ts

sindresorhus commented:
> It adds a checkbox to enable or disable the context menu, but for that to work while keeping it simple, I moved the entire logic our of the js and into the html file.

I think you should create a new fixture just for this. It's totally fine that it's in the HTML file. It can be as just `contextMenu()`, with no options. It should be part of this PR.

loopmode commented:
Great, will do that!

sindresorhus commented:
Bump :)

loopmode commented:
Hey @sindresorhus 

So.. I did quite some things extra here, hope it doesn't make accepting the PR too much of a problem?

- I added a .prettierrc which basically resembles the existing code style, at least I was able to format on save without causing extra changes
- I added a fixture for creating and disposing the menu, it can be used via `yarn start-fixture3`. I suggest you/we rename the fixture scripts :)
- Because the fixture files started to dominate the project root, i moved all of them into a `fixtures/` subfolder and adjusted the start-scripts accordingly
- The array of disposable functions does get emptied on dispose now, but I still did leave the `isDisposed` flag. As mentioned previously, it's currently essential. Maybe it's my lack of imagination, but I don't see another simple or elegant approach. Emptying the array and checking for length === 0 is unfortunately definitely not the way to go, as it cannot distinguish between "already disposed" and "not initialized yet". I'd be happy to learn something here, but really I think the boolean flag is the way to go

sindresorhus commented on readme.md:
````diff
@@ -52,10 +52,21 @@ let mainWindow;
 })();
 ```
 
+Note that the return value of `contextMenu()` is a dispose function:
````
The readme and index.d.ts should be in sync documentation-wise.

loopmode replied:
There are a couple of differences, as I didn't know about the sync aspect. Will adjust.
Actually, there's even a react hook example in index.d.ts that is not in readme. 
I will remove that altogether as it feels out of scope here.

loopmode replied:
> The readme and index.d.ts should be in sync documentation-wise.

Synced

sindresorhus commented on index.js:
```diff
@@ -217,14 +217,12 @@ const create = (win, options) => {
+				label: 'No Guesses Found',
+				visible: hasText && props.misspelledWord,
+				enabled: false
+			});
```
Don't do unrelated changes.

loopmode replied:
Yeah, sorry. Related to prettierrc - will disable auto-formatting in my editor.

loopmode replied:
> Don't do unrelated changes.

Reverted

sindresorhus commented:
> I added a .prettierrc which basically resembles the existing code style, at least I was able to format on save without causing extra changes

I'm not interested in having this.

sindresorhus commented:
> I added a fixture for creating and disposing the menu, it can be used via yarn start-fixture3. I suggest you/we rename the fixture scripts :)
Because the fixture files started to dominate the project root, i moved all of them into a fixtures/ subfolder and adjusted the package scripts accordingly
The array of disposable functions does get emptied on dispose now, but I still did leave the isDisposed flag. As mentioned previously, it's currently essential. Maybe it's my lack of imagination, but I don't see another simple or elegant approach. Emptying the array and checking for length === 0 is unfortunately definitely not the way to go, as it cannot distinguish between "already disposed" and "not initialized yet". I'd be happy to learn something here, but really I think the boolean flag is the way to go

👍

loopmode commented:
> > I added a .prettierrc which basically resembles the existing code style, at least I was able to format on save without causing extra changes
> 
> I'm not interested in having this.

removed

sindresorhus commented on package.json:
```diff
@@ -10,10 +10,12 @@
 		"email": "sindresorhus@gmail.com",
 		"url": "https://sindresorhus.com"
 	},
+	"contributors": ["Jovica Aleksic <jovica.aleksic@loopmode.de>"],
```
I don't use this field. It's not actually used anywhere and it's creates inequality with contributors that doesn't add themselves there. I prefer to keep it simple and just not use it. Hope you understand :)

loopmode replied:
Absolutely no problem! :)

sindresorhus commented:
Thanks :)
